src/JsonSchema/Validator.php

```diff
@@ -43,13 +41,45 @@ public function check($value, $schema = null, JsonPointer $path = null, $i = nul
      * types defined in the schema whenever possible. Note that the first argument is passed by reference,
      * so you must pass in a variable.
      *
+     * This method is now a simple alias, and exists only so that the API is not fundamentally broken for
+     * people who use this. This method should probably be officially deprecated at some point.
+     *
      * {@inheritDoc}
      */
-    public function coerce(&$value, $schema = null, JsonPointer $path = null, $i = null)
+    public function coerce(&$value, $schema = null)
     {
-        $validator = $this->factory->createInstanceFor('schema');
-        $validator->coerce($value, $schema);
+        // Preserve the current checkMode, because users of this API expect to be
+        // able to set it via the factory before calling this method.
+        $checkMode = $this->factory->getCheckMode() | self::CHECK_MODE_COERCE;
 
-        $this->addErrors(array_unique($validator->getErrors(), SORT_REGULAR));
+        return $this->validate($value, $schema, $checkMode);
+    }
+
+    /**
+     * Validates the given data against the schema and returns an object containing the results
+     * Both the php object and the schema are supposed to be a result of a json_decode call.
+     * The validation works as defined by the schema proposal in http://json-schema.org.
+     *
+     * @param mixed $value Decoded JSON document to validate
+     * @param mixed $schema Decoded JSON schema to validate against
+     * @param int $checkMode Set various options to change the validation behavior
+     *
+     * {@inheritDoc}
+     */
+    public function validate(&$value, $schema = null, $checkMode = self::CHECK_MODE_NORMAL)
+    {
+        $this->factory->setCheckMode($checkMode);
+        return $this->_check($value, $schema);
+    }
+
+    /**
+     * Alias of validate which does not allow any changes to the original document
+     * 
+     * {@inheritDoc}
+     */
+    public function validateImmutable($value, $schema = null, $checkMode = self::CHECK_MODE_NORMAL)
+    {
+        $checkMode &= $this->getCheckModeImmutableMask();
```

Comment:
It's there because there needs to be some kind of entry point for pass-by-value, and not enforcing immutability results in inconsistent behavior between object mode and assoc mode, because objects are *never* passed by value. Having the mask means that no matter what the user decides to feed it for $checkMode, the result will be the same regardless of whether $value is an object or an array.